### PR TITLE
Implement mix profile.tprof

### PR DESCRIPTION
**Pull request by sabiwara**
Close https://github.com/elixir-lang/elixir/issues/13581

Sharing my draft for feedback as I think the implementation is now done - I still need to do the moduledocs.

I used `profile.eprof` as a  base, since these are actually quite close.
Reviewing this commit [Actually implement tprof](https://github.com/elixir-lang/elixir/commit/5ac62df9013b26008964a7a11f9fe85bf1a70d33) might help the review since it highlights the differences.

<details>

<summary>Sample outputs</summary>

```elixir
mix profile.tprof -e "Enum.each(1..5, &String.Chars.Integer.to_string/1)"
Warmup...


Profile results of #PID<0.107.0>
#                                               CALLS      % TIME µS/CALL
Total                                              20 100.00    2    0.10
String.Chars.Integer.to_string/1                    5   0.00    0    0.00
anonymous fn/0 in :elixir_compiler_1.__FILE__/1     1   0.00    0    0.00
Enum.each/2                                         1   0.00    0    0.00
Enum.reduce_range/5                                 3   0.00    0    0.00
:erlang.integer_to_binary/1                         5  50.00    1    0.20
anonymous fn/3 in Enum.each/2                       5  50.00    1    0.20

Profile done over 6 matching functions
```

```elixir
mix profile.tprof -e "Enum.each(1..5, &String.Chars.Integer.to_string/1)" --type memory
Warmup...


Profile results of #PID<0.107.0>
#                           CALLS      % WORDS PER CALL
Total                           6 100.00    19     3.17
Enum.each/2                     1  21.05     4     4.00
:erlang.integer_to_binary/1     5  78.95    15     3.00

Profile done over 2 matching functions
```

```elixir
mix profile.tprof -e "Enum.each(1..5, &String.Chars.Integer.to_string/1)" --type calls
Warmup...


Profile results over all processes
#                                               CALLS      %
Total                                              20 100.00
anonymous fn/0 in :elixir_compiler_1.__FILE__/1     1   5.00
Enum.each/2                                         1   5.00
Enum.reduce_range/5                                 3  15.00
:erlang.integer_to_binary/1                         5  25.00
String.Chars.Integer.to_string/1                    5  25.00
anonymous fn/3 in Enum.each/2                       5  25.00

Profile done over 6 matching functions
```

```elixir
mix profile.tprof -e "Enum.each(1..5, &String.Chars.Integer.to_string/1)" --type calls --sort time
Warmup...

** (Mix) Incompatible sort option `time` with type `calls`
```

</details>

### CODE REVIEW

**Comment by sabiwara** [comment]
I still need to find a good conditional strategy to fix the CI warnings complaining about non-existing module.

Screenshot 2024-05-27 at 9 28 53

But this is a compile-time OTP version check, what we'd actually care for `tprof` existence is about is the runtime version.
Should we say that we compile a noop task if compiled with OTP26-?

Or use `apply` or some indirection to fool the compiler?

Will come back to it later.

File: lib/mix/lib/mix/tasks/profile.tprof.ex [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+  use Mix.Task
+
+  @shortdoc "Profiles the given file or expression with tprof"
+
+  @moduledoc """
+  Profiles the given file or expression using Erlang's `tprof` tool.
+
+  Requires Erlang/OTP27 or above.
+
+  [`:tprof`](`:tprof`) can provide time or memory information of each function call
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/profile.tprof.ex:11]:
```suggestion
  [`:tprof`](`:tprof`) can provide count, time, or memory information of each function call
```

File: lib/mix/lib/mix/tasks/profile.tprof.ex [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+  and this may need special attention, when: running asynchronous code as function calls which were
+  called before the profiler stopped will not be counted; running synchronous code as long
+  running computations and a profiler without a proper MFA trace pattern or filter may
+  lead to a result set which is difficult to comprehend.
+
+  You should expect a slowdown in your code execution using this tool since `:tprof` has
+  some performance impact on the execution, but the impact is considerably lower than
+  `Mix.Tasks.Profile.Fprof`. If you have a large system try to profile a limited
+  scenario or focus on the main modules or processes. Another alternative is to use
+  `Mix.Tasks.Profile.Cprof` that uses [`:cprof`](`:cprof`) and has a low performance degradation effect.
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/profile.tprof.ex:110]:
```suggestion
  scenario or focus on the main modules or processes.
```

This is pretty much the same as cprof (which will be deprecated in the future anyway).

File: lib/mix/lib/mix/tasks/profile.tprof.ex [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+    return_value
+  end
+
+  defp profile_and_analyse(fun, opts) do
+    if Keyword.get(opts, :warmup, true) do
+      IO.puts("Warmup...\n")
+      fun.()
+    end
+
+    tprof_module().start()
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/profile.tprof.ex:230]: I don't think we need to start and stop it, do we?

File: lib/mix/lib/mix/tasks/profile.tprof.ex [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+    measurement_opt =
+      get_filter_value!(type, Keyword.get(opts, :time), Keyword.get(opts, :memory))
+
+    Enum.filter(call_results, fn {_module, _fa, count, measurement, _, _} ->
+      count >= calls_opt and measurement >= measurement_opt
+    end)
+  end
+
+  defp get_filter_value!(type, time, _memory) when is_integer(time) and type != :time do
+    Mix.raise("Incompatible use of time option with type `#{type}`")
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/profile.tprof.ex:287]: We don't use backticks in Elixir error messages. Perhaps #{inspect(type)} is enough?

**Review by josevalim — APPROVED** [review]
Beautifully done!

File: lib/mix/lib/mix/tasks/profile.tprof.ex (hunk 2 above)
Comment (sabiwara) [reply lib/mix/lib/mix/tasks/profile.tprof.ex:110]: Yes sorry as I mentioned in the description I should still rewrite the bulk of the moduledoc, it's still mostly `eprof`'s 😅

File: lib/mix/lib/mix/tasks/profile.tprof.ex (hunk 4 above)
Comment (sabiwara) [reply lib/mix/lib/mix/tasks/profile.tprof.ex:287]: Sounds much better indeed. I needed some kind of visual distinction to keep readability but atom notation feels much more natural.

File: lib/mix/lib/mix/tasks/profile.tprof.ex (hunk 3 above)
Comment (sabiwara) [reply lib/mix/lib/mix/tasks/profile.tprof.ex]: You're right we don't [27e56dc](https://github.com/elixir-lang/elixir/pull/13605/commits/27e56dc01ae5ff6f97c5d175237515bd2226ff4c)

File: lib/mix/lib/mix/tasks/profile.tprof.ex (hunk 4 above)
Comment (sabiwara) [reply lib/mix/lib/mix/tasks/profile.tprof.ex]: [27e56dc](https://github.com/elixir-lang/elixir/pull/13605/commits/27e56dc01ae5ff6f97c5d175237515bd2226ff4c)

**Comment by sabiwara** [comment]
OK I think we should now be good. Updated the moduledoc https://github.com/elixir-lang/elixir/pull/13605/commits/adf24b64a1cfafd1433d074e8fcc8a44bf724bdc.
Reused some of the wordings from the [original doc](https://www.erlang.org/doc/apps/tools/tprof.html):
Screenshot 2024-05-27 at 21 38 02

**Review by josevalim — APPROVED** [review]
The whole doc looked good to me before and it still looks good to me now :D

**Pull request merged**